x/ccv/provider/keeper/proposal.go

```diff
@@ -29,6 +30,75 @@ func (k Keeper) CreateConsumerChainProposal(ctx sdk.Context, p *types.CreateCons
 	return nil
 }
 
+// StopConsumerChainProposal stops a consumer chain and released the outstanding unbonding operations.
+// If the stop time hasn't already passed, it stores the proposal as a pending proposal.
+func (k Keeper) StopConsumerChainProposal(ctx sdk.Context, p *types.StopConsumerChainProposal) error {
+	if ctx.BlockTime().After(p.StopTime) {
+		return k.StopConsumerChain(ctx, p.ChainId, false)
+	}
+
+	k.Hooks().k.SetPendingStopProposal(ctx, p.ChainId, p.StopTime)
+	return nil
+}
+
+// StopConsumerChain cleans up the states for the given consumer chain ID and, if the given lockUbd is false,
+// it completes the outstanding unbonding operations lock by the consumer chain.
+func (k Keeper) StopConsumerChain(ctx sdk.Context, chainID string, lockUbd bool) (err error) {
+
+	// clean up states
+	k.DeleteConsumerClient(ctx, chainID)
+	k.DeleteLockUnbondingOnTimeout(ctx, chainID)
+
+	// close channel and delete the mappings between chain ID and channel ID
+	if channelID, found := k.GetChainToChannel(ctx, chainID); found {
+		k.chanCloseInit(ctx, channelID)
```

Comment:
I'm not sure whether this is correct (I know that it's following the spec, but it may be wrong there also). I'm thinking of the logic of the closing handshake. This function is called by the provider as a result of either a gov proposal or a timeout. 

- In the case of a gov proposal, clearly we want to close this channel end (by calling `chanCloseInit`), so a relayer can relay a `MsgChannelCloseConfirm` to the consumer. 
- In the case of a timeout, this function is called from `OnTimeoutPacket()`, which is called on receiving either `MsgTimeout` (i.e., a packet sent by the provider timed out) or `MsgTimeoutOnClose` (i.e., a packet sent by the provider cannot be received since the consumer's channel end is closed). In both cases, the provider's channel end will be closed once either of the packets is handled (see [here](https://github.com/cosmos/ibc-go/blob/b2ca1932849cc3f05fe9e6ec30049aa4d08e6fc6/modules/core/04-channel/keeper/timeout.go#L157)).  Thus, we shouldn't call `chanCloseInit` in this case. 

@AdityaSripal Do you agree with this?

The consumer could also call `chanCloseInit` when a packet send to the provider times out. This will solve the following edge case, which is very unlikely:
- a packet send to the provider times out
- the consumer's channel end closes 
- the provider has no validator updates to send to the consumer, thus there is no packet for which a relayer can send a `MsgTimeoutOnClose`, which means that the provider never remove the consumer chain. 

Not sure whether it's worth the time to cover this edge case, since eventually there will be a validator set update on the provider, which will timeout  and lead to the call of `StopConsumerChain`.

Reply:
Also, as a sanity check, we should just check if the channel is already closed before calling `chanCloseInit`